---
title: "mouse"
author: "Naoki Matsumaru"
date: "2024-04-17"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r reading libraries}
library(tidyverse)
library(Seurat)
library(openxlsx)
library(hdf5r)
library(ggplot2)
library(HGNChelper)
#library(SeuratData)
set.seed(1234)
```
Data acquisition

source:

https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE224398

available on h5 format.

```{r reading h5 files}
h5_filtered_WT <- Read10X_h5("GSM7021714_C_6m_filtered_feature_bc_matrix.h5")
h5_filtered_AD <- Read10X_h5("GSM7021717_A_6m_filtered_feature_bc_matrix.h5")
```


```{r eval=FALSE}
h5_filtered_WT %>% class
h5_filtered_WT %>% dim
h5_filtered_WT %>% as.data.frame()
```

dgcMatrix is a class of sparse numeric matrices in the compressed, sparse, column-oriented format.


```{r}

seurat_object_WT <- CreateSeuratObject(counts = h5_filtered_WT, project = "ScRNA-seq of AD mouse", min.cells = 3, min.features = 200)

seurat_object_AD <- CreateSeuratObject(counts = h5_filtered_AD, project = "ScRNA-seq of AD mouse", min.cells = 3, min.features = 200)

```
```{r}
seurat_object_WT$type <- "WT"
seurat_object_AD$type<- "AD"

```

```{r}
seurat_object_WT@meta.data
seurat_object_WT %>% names
seurat_object_WT[["RNA"]]
```

```{r}
h5_filtered_WT %>% as.data.frame() %>% filter(str_detect(row.names(.), "^MT-"))
h5_filtered_AD %>% as.data.frame() %>% filter(str_detect(row.names(.), "^MT-"))
```

No MT related genes

```{r}
###QUALITY CHECK###-------------------------------------------------------------
seurat_object_WT[["percent.mt"]] <- PercentageFeatureSet(seurat_WT, pattern = "^MT-")
seurat_object_AD[["percent.mt"]] <- PercentageFeatureSet(seurat_AD, pattern = "^MT-")

VlnPlot(seurat_WT, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
#FeatureScatter(seurat_WT, feature1 = "nCount_RNA", feature2 = "percent.mt")
FeatureScatter(seurat_WT, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
```
```{r}
VlnPlot(seurat_AD, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
#FeatureScatter(seurat_AD, feature1 = "nCount_RNA", feature2 = "percent.mt")
FeatureScatter(seurat_AD, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
```


```{r}
seurat_WT <- subset(seurat_WT, subset = nFeature_RNA > 200 & nFeature_RNA < 5000 & percent.mt < 30)
seurat_AD <- subset(seurat_AD, subset = nFeature_RNA > 200 & nFeature_RNA < 3000 & percent.mt < 25)
```


```{r}
###NORMALIZATION###-------------------------------------------------------------
seurat_WT <- NormalizeData(seurat_WT, normalization.method = "LogNormalize", scale.factor = 10000)
seurat_AD <- NormalizeData(seurat_AD, normalization.method = "LogNormalize", scale.factor = 10000)
```


```{r}
seurat_WT <- FindVariableFeatures(seurat_WT, selection.method = "vst", nfeatures = 3000)
seurat_AD <- FindVariableFeatures(seurat_AD, selection.method = "vst", nfeatures = 3000)
```


```{r, fig.width=13, fig.height=5}
top10 <- head(VariableFeatures(seurat_WT), 10)
plot1 <- VariableFeaturePlot(seurat_WT)
plot21 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
#plot21

top10 <- head(VariableFeatures(seurat_AD), 10)
plot1 <- VariableFeaturePlot(seurat_AD)
plot22 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot21 + plot22
```

# Data scaling

```{r}
seurat_WT <- ScaleData(seurat_WT)
seurat_WT <- RunPCA(seurat_WT, features = head(VariableFeatures(object=seurat_WT), 10000))
DimHeatmap(seurat_WT, dims = 1:3, cells = 500, balanced = T)

seurat_AD <- ScaleData(seurat_AD)
seurat_AD <- RunPCA(seurat_AD, features = head(VariableFeatures(object=seurat_AD), 10000))
DimHeatmap(seurat_AD, dims = 1:3, cells = 500, balanced = T)

```


```{r}
source(file.path(getwd(),"..", "..", "sc-type", "R", "gene_sets_prepare.R"))
source(file.path(getwd(),"..", "..", "sc-type", "R", "sctype_score_.R"))
```


```{r}
db_ = "SCTypeDB_full.xlsx"
tissue = "Brain"
```

```{r}
gs_list = gene_sets_prepare(db_, tissue)
```

```{r}
scRNAseqData = readRDS(file.path(getwd(), "..","..", "sc-type", "exampleData.RDS")); 
es.max = sctype_score(scRNAseqData = scRNAseqData, scaled = TRUE, gs = gs_list$gs_positive, gs2 = gs_list$gs_negative)
```


```{r}
#integration of Data
mouse_all.list <- NULL
mouse_all.list <- list(seurat_WT,seurat_AD)

names(mouse_all.list) <- c("WT", "AD")


###Áµ±Âê?###----------------------------------------------------------------------
mouse_all.anchors <- FindIntegrationAnchors(object.list = mouse_all.list, dims = 1:20)
mouse_all.integrated <- IntegrateData(anchorset = mouse_all.anchors, dims = 1:20)
DefaultAssay(mouse_all.integrated) <- "integrated"
```

```{r}
###PCA###-----------------------------------------------------------------------
mouse_all.integrated <- ScaleData(mouse_all.integrated, verbose=FALSE)
mouse_all.integrated <- RunPCA(mouse_all.integrated, npcs=30, verbose = FALSE)
```

‰ªÆ„Å´PC1-20„ÅßËß£Êûê„ÇíÈÄ≤„ÇÅ„Å¶„Å?„Åæ„Å?

```{r}
mouse_all.integrated <- RunUMAP(mouse_all.integrated, reduction = "pca", dims = 1:20)

###CLUTERING###-----------------------------------------------------------------
mouse_all.integrated <- FindNeighbors(mouse_all.integrated, reduction = "pca", dims = 1:20)
mouse_all.integrated <- FindClusters(mouse_all.integrated, resolution = 0.5)
```


```{r}
#saveRDS(merge.data,"~/Desktop/Save/WT:AD_Save/mergeData_donePCA.rds")
DimPlot(mouse_all.integrated, reduction = "umap", split.by = "type", label = TRUE)
```
```{r, fig.height=5, fig.width=6}
DimPlot(mouse_all.integrated, reduction = "umap", group.by="type", label=TRUE)
```

#Differentially expressed gene (DEG)„ÅÆÊ§úÂ?∫
```{r}
mouse_AD.markers <- FindAllMarkers(mouse_all.integrated, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
```

```{r}
mouse_AD.markers %>% group_by(cluster) %>% top_n(n=2, wt = avg_log2FC)
```
```{r}
top10 <- mouse_AD.markers %>% group_by(cluster) %>% top_n(n=10, wt=avg_log2FC)
```

```{r}
cluster1.markers <- FindMarkers(mouse_all.integrated, ident.1=1, min.pct = 0.25)
cluster5.markers <- FindMarkers(mouse_all.integrated, ident.1=5, ident=c(0,3), min.pct=0.25)
```
```{r}
mouse_all.integrated$type %>% tail
```


```{r, fig.height=20, fig.width=20}
DoHeatmap(mouse_all.integrated, features=top10$gene)
```


```{r}
mouse_all.integrated %>% names

mouse_all.integrated@meta.data
mouse_all.integrated@
```


```{r}
mouse.AD.marker <- FindMarkers(mouse_all.integrated, ident.1 = "WT", ident.2 = "AD")
```


```{r}

###ANNOTATION###----------------------------------------------------------------

#ÈôΩÊÄßÈÅ∫‰ºùÂ≠ê„Å®Èô∞ÊÄßÈÅ∫‰ºùÂ≠ê„?Æ„Çπ„Ç≥„Ç¢„É™„É≥„Ç∞
es.max = sctype_score(scRNAseqData = mouse_all[["integrated"]]@scale.data, scaled = TRUE, gs = gs_list$gs_positive, gs2 = gs_list$gs_negative)
#meta.data„ÅÆseurat_clusters„Çì„ÇÇÊÉ?Â†±„Å®es.max„ÅÆÊÉ?Â†±„ÇíÁµ±Âê?
cL_resutls = do.call("rbind", lapply(unique(merge.data@meta.data$seurat_clusters), function(cl){
  es.max.cl = 
sort(rowSums(es.max[ ,rownames(merge.data@meta.data[merge.data@meta.data$seurat_clusters==cl, ])]), decreasing = !0)
  head(data.frame(cluster = cl, type = names(es.max.cl), scores = es.max.cl, ncells = sum(merge.data@meta.data$seurat_clusters==cl)), 10)
}))
  sctype_scores = cL_resutls %>% group_by(cluster) %>% top_n(n = 1, wt = scores) 
#‰ø°È†ºÊÄß„ÅÆ‰Ωé„ÅÑsctype_score„ÅÆÂ†¥Âêà„Åù„ÅÆ„ÇØ„É©„Çπ„Çø„Éº„Çíunknown„Å´
  sctype_scores$type[as.numeric(as.character(sctype_scores$scores)) < 
sctype_scores$ncells/4] = "unknown"
  print(sctype_scores[,1:3])
#Á¥∞ËÉûË?´„ÇíÁâπÂÆö„Åô„ÇãÊñ∞„Åü„Å™meta.data„Åß„ÅÇ„Çã"customclassif"„ÇíËøΩÂä?„Åô„Çã
merge.data@meta.data$customclassif = ""
for (j in unique(sctype_scores$cluster)) {
  cl_type <- sctype_scores[sctype_scores$cluster == j,]
  merge.data@meta.data$customclassif[merge.data@meta.data$seurat_clusters == j] <- as.character(cl_type$type[1])
}

DimPlot(merge.data, reduction = "umap", split.by = "type", label = TRUE, repel = TRUE, group.by = 'customclassif') 

```

## R Markdown


```{r eval=FALSE}
###DATASET###-------------------------------------------------------------------
setwd("~/Desktop/R_Seurat/AD.scRNA_Seq/")

source("https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/R/gene_sets_prepare.R"); 
source("https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/R/sctype_score_.R")
db_ = "https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/ScTypeDB_full.xlsx";
tissue = "Brain" 

gs_list = gene_sets_prepare(db_, tissue)
scRNAseqData = readRDS(gzcon(url('https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/exampleData.RDS'))); 
es.max = sctype_score(scRNAseqData = scRNAseqData, scaled = TRUE, gs = gs_list$gs_positive, gs2 = gs_list$gs_negative)

raw_WT <- Read10X_h5("~/Desktop/Downloaded.Data/AD:WT/GSM7021714_C_6m_filtered_feature_bc_matrix.h5")
raw_AD <- Read10X_h5("~/Desktop/Downloaded.Data/AD:WT/GSM7021717_A_6m_filtered_feature_bc_matrix.h5")

raw_WT <- CreateSeuratObject(counts = raw_WT, project = "WT", min.cells = 3, min.features = 200)
raw_WT$type <- "WT"

raw_AD <- CreateSeuratObject(counts = raw_AD, project = "AD", min.cells = 3, min.features = 200)
raw_AD$type <- "AD"

###QUALITY CHECK###-------------------------------------------------------------
raw_WT[["percent.mt"]] <- PercentageFeatureSet(raw_WT, pattern = "^MT-")
raw_AD[["percent.mt"]] <- PercentageFeatureSet(raw_AD, pattern = "^MT-")

VlnPlot(raw_WT, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
FeatureScatter(raw_WT, feature1 = "nCount_RNA", feature2 = "percent.mt")
FeatureScatter(raw_WT, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")

VlnPlot(raw_AD, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
FeatureScatter(raw_AD, feature1 = "nCount_RNA", feature2 = "percent.mt")
FeatureScatter(raw_AD, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")

raw_WT <- subset(raw_WT, subset = nFeature_RNA > 200 & nFeature_RNA < 5000 & percent.mt < 30)
raw_AD <- subset(raw_AD, subset = nFeature_RNA > 200 & nFeature_RNA < 3000 & percent.mt < 25)

###NORMALIZATION###-------------------------------------------------------------
raw_WT <- NormalizeData(raw_WT, normalization.method = "LogNormalize", scale.factor = 10000)
raw_AD <- NormalizeData(raw_AD, normalization.method = "LogNormalize", scale.factor = 10000)

raw_WT <- FindVariableFeatures(raw_WT, selection.method = "vst", nfeatures = 3000)
raw_AD <- FindVariableFeatures(raw_AD, selection.method = "vst", nfeatures = 3000)

top10 <- head(VariableFeatures(raw_WT), 10)
plot1 <- VariableFeaturePlot(raw_WT)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot1 + plot2

top10 <- head(VariableFeatures(raw_AD), 10)
plot1 <- VariableFeaturePlot(raw_AD)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot1 + plot2

DoHeatmap(merge.data, dims = 1:3, cells = 500, balanced = T)
###Áµ±Âê?###----------------------------------------------------------------------
merge.anchors <- FindIntegrationAnchors(object.list = list(raw_WT, raw_AD), dims = 1:20)
merge.data <- IntegrateData(anchorset = merge.anchors, dims = 1:20)
DefaultAssay(merge.data) <- "integrated"

###PCA###-----------------------------------------------------------------------
merge.data <- ScaleData(merge.data, verbose = FALSE)
merge.data <- RunPCA(merge.data, npcs = 30, verbose = FALSE)

###CLUTERING###-----------------------------------------------------------------
merge.data <- FindNeighbors(merge.data, reduction = "pca", dims = 1:20)
merge.data <- FindClusters(merge.data, resolution = 0.5)
merge.data <- RunUMAP(merge.data, reduction = "pca", dims = 1:20)

saveRDS(merge.data,"~/Desktop/Save/WT:AD_Save/mergeData_donePCA.rds")
DimPlot(merge.data, reduction = "umap", split.by = "type", label = TRUE)

###ANNOTATION###----------------------------------------------------------------
merge.data <- readRDS("~/Desktop/Save/WT:AD_Save/mergeData_donePCA.rds")
#ÈôΩÊÄßÈÅ∫‰ºùÂ≠ê„Å®Èô∞ÊÄßÈÅ∫‰ºùÂ≠ê„?Æ„Çπ„Ç≥„Ç¢„É™„É≥„Ç∞
es.max = sctype_score(scRNAseqData = merge.data[["integrated"]]@scale.data, scaled = TRUE, gs = gs_list$gs_positive, gs2 = gs_list$gs_negative)
#meta.data„ÅÆseurat_clusters„Çì„ÇÇÊÉ?Â†±„Å®es.max„ÅÆÊÉ?Â†±„ÇíÁµ±Âê?
cL_resutls = do.call("rbind", lapply(unique(merge.data@meta.data$seurat_clusters), function(cl){
  es.max.cl = 
sort(rowSums(es.max[ ,rownames(merge.data@meta.data[merge.data@meta.data$seurat_clusters==cl, ])]), decreasing = !0)
  head(data.frame(cluster = cl, type = names(es.max.cl), scores = es.max.cl, ncells = sum(merge.data@meta.data$seurat_clusters==cl)), 10)
}))
  sctype_scores = cL_resutls %>% group_by(cluster) %>% top_n(n = 1, wt = scores) 
#‰ø°È†ºÊÄß„ÅÆ‰Ωé„ÅÑsctype_score„ÅÆÂ†¥Âêà„Åù„ÅÆ„ÇØ„É©„Çπ„Çø„Éº„Çíunknown„Å´
  sctype_scores$type[as.numeric(as.character(sctype_scores$scores)) < 
sctype_scores$ncells/4] = "unknown"
  print(sctype_scores[,1:3])
#Á¥∞ËÉûË?´„ÇíÁâπÂÆö„Åô„ÇãÊñ∞„Åü„Å™meta.data„Åß„ÅÇ„Çã"customclassif"„ÇíËøΩÂä?„Åô„Çã
merge.data@meta.data$customclassif = ""
for (j in unique(sctype_scores$cluster)) {
  cl_type <- sctype_scores[sctype_scores$cluster == j,]
  merge.data@meta.data$customclassif[merge.data@meta.data$seurat_clusters == j] <- as.character(cl_type$type[1])
}

DimPlot(merge.data, reduction = "umap", split.by = "type", label = TRUE, repel = TRUE, group.by = 'customclassif') 

###„Ç¢„Éé„ÉÜ„Éº„Ç∑„Éß„É≥„Åó„ÅüÁ¥∞ËÉûË?´„ÇíÊäú„ÅçÂ?∫„Åó„Å¶„ÄÅDEG„ÇíË¶ã„Å§„Åë„Çã###-----------------------
Idents(merge.data) <- "customclassif"
#merge.data2 <- subset(merge.data, idents = 0)
#Dimplot„ÅßÁ¢∫Ë™ç„Åß„Åç„Çã
MC_data <- subset(merge.data, idents = "Microglial cells")
#MC_data <- SetIdent(MC_data, value = "type")
MC_list<- SplitObject(object = MC_data, split.by = "type")
Galectin <- subset(merge.data, idents = "LGALS3")
#DEG„ÅÆÊé¢Á¥¢„Å®pÂÄ§„ÅÆFDR„Å´„Çà„ÇãBHÊ≥ïË£úÊ≠£
MC_marker <- FindMarkers(MC_data, ident.1 = "WT", assay = "RNA", lfcThreshold = 0, min.cells.feature = 1, min.cells.group = 1, min.pct = 0)
MC_marker$p_val_adj = p.adjust(MC_marker$p_val, method = 'fdr')
write.csv(MC_marker, "RNAMC_DEG.csv")
VlnPlot(MC_data, "Galectin3", assay = "RNA")

DoHeatmap(object = MC_data, group.by = "type", slot = "scale.data")
###merge.data„ÇíWT„Å®AD„ÅßÂà?Ââ≤„Åó„Å¶listÂåñ„Åô„Ç?###------------------------------------
type_data <- SplitObject(merge.data, split.by = "type")
#tmp„Å®„ÅØtemporary„ÅÆ„Åì„Å®„ÄÅÊÑèÂë≥„ÅÆ„Å™„Å?‰∏ÄÊôÇÁöÑ„Å™„Éï„Ç°„Ç§„É´
#table()„Çí‰Ωø„Å?„Å®Âê?„Ç§„É≥„É?„É≥„Éà„?ÆÂõûÊï∞„Ç?"Freq"„Å®„Åó„Å¶‰øùÂ≠ò„Ä?
#Var1,Var2Âàó„ÇÇËá™ÂãïÁöÑ„ÅßÁîüÊ?ê„Åï„Çå„Çã„ÄÇ„Åù„Çå„Åû„Ç?"ident"„Å®"type"„ÅåÊ?ºÁ¥ç„Åï„Çå„Å¶„Å?„Ç?
tmp1 <- as.data.frame(table(Idents(type_data$WT), type_data$WT$type))
tmp2 <- as.data.frame(table(Idents(type_data$AD), type_data$AD$type))
#tmp„Çíbind_rows()„ÅßÁµêÂêà„Åï„Åõ„Ç?
pt <- bind_rows(tmp1, tmp2)
pt$Var1 <- as.character(pt$Var1)

ggplot(pt, aes(x = Var2, y = Freq, fill = Var1)) +
  theme_bw(base_size = 15) +
  geom_col(position = "fill", width = 0.5) +
  xlab("Sample") +
  ylab("Proportion") +
  theme(legend.title = element_blank())

MC_data <- subset(merge.data, idents = "Microglial cells")
MC_data <- SetIdent(MC_data, value = "type")

###Âê?„ÇØ„É©„Çπ„Çø„Éº„ÅÆDEG„ÅÆ„Ç®„É≥„É™„É?„ÉÅËß£Êûê„Å®ÊØîËº?###-----------------------------------
library(dplyr)
library(clusterProfiler)
library(org.Hs.eg.db)

deg <- FindAllMarkers(object = merge.data, only.pos = T)
table(deg$cluster)
Seurat::DoHeatmap(raw_AD, features = gene_list,)
#„ÇØ„É©„Çπ„Çø„Éº„Åî„Å®„ÅÆDEG„Çí„?ô„ÇØ„Éà„É´„Å´„Åó„Å¶„É™„Çπ„Éà„Å´„Åæ„Å®„ÇÅ„Çã
gene_list <- split(deg$gene, deg$cluster)
#ÈÅ∫‰ºùÂ≠êÂêç„Åã„ÇâÈÅ∫‰ºùÂ≠êID„Å´Â§âÊèõ
for(i in names(gene_list)){
  gene_list[[i]] <- bitr(gene_list[[i]],
                           fromType = "SYMBOL",
                           toType = "ENTREZID",
                           OrgDb = "org.Hs.eg.db")
                           }
#ÊÆã„Å£„ÅüÈÅ∫‰ºùÂ≠ê„ÇíË™ø„Åπ„Ç?
length(unlist(gene_list))/2

#clusterProfilerÁî®„Å´ÊàêÂΩ¢
for(i in names(gene_list)){
  gene_list[[i]] <- gene_list[[i]]$ENTREZID
}

#Gene Ontology
GOBP <- compareCluster(geneClusters = gene_list,
                          fun = "enrichGO",
                          ont = "BP",
                          OrgDb = "org.Hs.eg.db")

#KEGG
kegg <- compareCluster(geneClusters = gene_list,
                       fun = "enrichKEGG",
                       organism = "mmu")

#Reactome
library(ReactomePA)
reactome <- compareCluster(geneClusters = gene_list,
                           fun = "enrichPathway",
                           organism = "mouse")

#WikiPathway
WP_clusterplot <- compareCluster(geneClusters = gene_list,
                                 fun = "enrichWP",
                                 organism = "Mus musculus")

#Disease Ontology
library(DOSE)
DO <- compareCluster(geneClusters = gene_list,
                     fun = "enrichDO")
#DisGeNET
DGN <- compareCluster(geneClusters = gene_list,
                      fun = "enrichDGN")
#ÁµêÊûú
dotplot(GOBP)


###volcanoplot###---------------------------------------------------------------
result <- readRDS("~/Desktop/Save/WT:AD_Save/mergeData_donePCA.rds")
selectLab <- result %>% filter(PValue < 0.05) %>% filter(abs(logFC) > 2.5) %>% .$extemal_gene_name


FeaturePlot(object = merge.data, features = "Acta2", label = TRUE, reduction = "umap", split.by = "type", group.by = 'customclassif')

```

